tools/maphash/lib/dmm.js

```diff
@@ -0,0 +1,471 @@
+				yield String.fromCharCode(i);
+			}
+			return;
+		}
+		for(let first_char of this.potential_keys(1)) {
+			for(let the_rest of this.potential_keys(key_len - 1)) {
+				yield first_char + the_rest;
+			}
+		}
+	}
+
+	number_to_key(num, key_len = 3) {
+		let arr = [];
+		for(let i = 0; i < key_len; i++) {
+			let num_thing = (num % 52);
+			if(num_thing < 26) arr.push(String.fromCharCode(97 + num_thing));
+			else arr.push(String.fromCharCode(65-26+num_thing));
+			num = Math.floor(num / 52);
+		}
+		return arr.reverse().join("");
+	}
+
+	toString() {
+		let endl = this.line_endings;
+		let out = "// This map has been converted using maphash" + endl;
+		let instance_strings = new Map();
+		for(let [x,y,z] of this.all_coordinates()) {
+			let str = this.get_tile(x,y,z).join(",");
+			instance_strings.set(str, (instance_strings.get(str) || 0) + 1);
+		}
+		let key_len = 1;
+		while(instance_strings.size > 52**key_len)
+			key_len++;
+		let used_keys = new Set();
+		let instance_keys = new Map();
+		
+		let collision_arr = [];
+		let instance_string_comparator = (a,b) => {
+			// Instances with more occurences get higher precedence.
+			// This is to prevent situations where a hash conflict causes the hash of something like
+			// (/turf/open/space/basic,/area/space) occuring and re-keying the entire space area.
+			if(a[1] == b[1]) return b[1] - a[1];
+			return a[0] > b[0] ? 1 : (a[0] < b[0] ? -1 : 0);
+		};
+		for(let [str] of [...instance_strings].sort(instance_string_comparator)) {
+			let key;
+			let i;
+			for(i = 0; i < 10000; i++) {
+				let hash_num = parseInt(crypto.createHash('sha1').update(str + i, 'utf8').digest('hex').substring(0, 4), 16);
+				let hash_key = this.number_to_key(hash_num);
+				if(!used_keys.has(hash_key) && hash_num < 65535) {
+					key = hash_key;
+					break;
+				}
+			}
+			used_keys.add(key);
+			collision_arr[i] = (collision_arr[i] || 0) + 1;
+			if(!key) throw new Error('Could not find a key');
+			instance_keys.set(str, key);
+		}
+		for(let [str, key] of [...instance_keys].sort((a,b) => {return compare_keys(a[1],b[1]);})) {
+			out += `"${key}" = (${str})${endl}`;
+		}
+		this.collision_stats = collision_arr;
+		out += endl;
+		for(let z = 1; z <= this.maxz; z++) {
+			for(let x = 1; x <= this.maxx; x++) {
+				out += `(${x},1,${z}) = {"${endl}`;
+				for(let y = this.maxy; y >= 1; y--) {
+					out += instance_keys.get(this.get_tile(x,y,z).join(","));
+					out += `${endl}`;
+				}
+				out += `"}${endl}`;
+			}
+			if(z != this.maxz) out += endl;
+		}
+		return out;
+	}
+}
+
+function key_to_number(key) {
+	let num = 0;
+	for(let i = 0; i < key.length; i++) {
+		num *= 52;
+		let c = key[i];
+		if(c >= 'a' && c <= 'z') {
+			num += (c.charCodeAt(0) - 'a'.charCodeAt(0));
+		} else {
+			num += (c.charCodeAt(0) - 'A'.charCodeAt(0)) + 26;
+		}
+	}
+	return num;
+}
+
+function compare_keys(a, b) {
+	return key_to_number(a) - key_to_number(b);
+}
+DMM.Instance = Instance;
+
+module.exports = DMM;
```

Comment:
I have absolutly not idea how this file works and i lack the motivation to figure what it does